### DISCUSSION TITLE
Installation stuck on errors

### DISCUSSION THREAD

**Discussion by emilytweaks** [Homebrew/discussions#4711] [Getting started]
### Output of `brew config`

```shell
HOMEBREW_VERSION: 4.1.4
ORIGIN: https://github.com/Homebrew/brew
HEAD: 98f1be77f93a0d23fbc68629d0854522e9f0c1c3
Last commit: 4 days ago
Core tap JSON: 10 Aug 04:29 UTC
HOMEBREW_PREFIX: /opt/homebrew
HOMEBREW_CASK_OPTS: []
HOMEBREW_MAKE_JOBS: 8
Homebrew Ruby: 2.6.10 => /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/bin/ruby
CPU: octa-core 64-bit arm_blizzard_avalanche
Clang: 14.0.3 build 1403
Git: 2.39.2 => /Library/Developer/CommandLineTools/usr/bin/git
Curl: 8.1.2 => /usr/bin/curl
macOS: 13.5-arm64
CLT: 14.3.1.0.1.1683849156
Xcode: N/A
Rosetta 2: false
```

### Output of `brew doctor`

```shell
Your system is ready to brew.
```

### Description of issue

Hello everyone!  Recently got a new Macbook Pro with Ventura 13.5 on it.  Always used macs, but in recent years it was annoying/tricky getting it to still let me use my darn Telnet for certain things- although I'm not a programmer, I could follow enough lines in the recommended Homebrew to discover how to let me install/port in.  This time, though...

Installed Homebrew successfully, it mentioned various things it was going to install/have to create new directories for.  Then it continued to doing all the command line tools for Xcode.  After which showed a huge list of objects it made, then my first warning about how '/opt/homebrew/bin is not in your PATH.'

==> Next steps:
- Run these two commands in your terminal to add Homebrew to your PATH:
    (echo; echo 'eval "$(/opt/homebrew/bin/brew shellenv)"') >> /Users/emilycollins/.profile
    eval "$(/opt/homebrew/bin/brew shellenv)"
- Run brew help to get started
- Further documentation:
    https://docs.brew.sh

Ok, I'll look into that... but first it auto-updated, then said:

You have 15 outdated formulae installed.

/usr/local/Homebrew/Library/Homebrew/brew.rb:19:in `fetch': key not found: "HOMEBREW_REQUIRED_RUBY_VERSION" (KeyError)
Did you mean?  "HOMEBREW_USER_AGENT_CURL"
	from /usr/local/Homebrew/Library/Homebrew/brew.rb:19:in `<main>'

I figure those errors are related to the paths/directory and do first, so I tried copying each of the two individual lines up there it told me to run first, and it says nothing in response.

So, um....  what next?  Do I need to update something?  Am I missing some ridiculously simple command, such as needing to type 'run' or 'command' before those two 'next steps' lines up there?  Those are all the things I can think of to fix, along with the false Rosetta thing... Sorry for the verbosity, was just trying to be thorough.  Thank you for any help!

**Comment by ZhongRuoyu** [comment]
It seems that you have restored the Homebrew install from your old Mac (via Time Machine or similar) so now you have two Homebrew installations. Below are steps to migrate the old installation to the one you've just installed.

Firstly, dump the list of previously-installed packages:

```
/usr/local/bin/brew bundle dump
```

Then, ask the new Homebrew to install them:

```
/opt/homebrew/bin/brew bundle --file Brewfile
```

(Or, if you don't care about the previously-installed packages, skip the `brew bundle` steps above.)

After that, remove the old Homebrew installation with:

```
curl -fsSLO https://raw.githubusercontent.com/Homebrew/install/HEAD/uninstall.sh
/bin/bash ./uninstall.sh --path /usr/local
```

You may want to have a look at `~/.zprofile` (or `~/.profile`) and clean up the configs for the removed installation (specifically, remove `eval "$(/usr/local/bin/brew shellenv)"`).

More discussions on migration can be found at https://github.com/orgs/Homebrew/discussions/417.

**Comment by emilytweaks** [comment]
Thank you, that was very helpful.  Makes sense due to using Migration Assistant, I had just assumed something no longer worked (or was lost) with the new system due to no longer being able to port to the server I wanted to in Telnet.  The other thread was also helpful, although I feel quite like one poster talking about how we can figure out typing in the commands we're told, and what happens, but we don't know what to do with it afterwards... sorry, I'm not used to coding much!  I understand them with talking about the difference in systems and making sure which one I'm in/directory/path, but not sure what to do after that, or checking that mine's correct now!

I followed each of your command instructions... got a couple errors about shallow clone and straggler files (which it then gave me the commands to also fix).  After all of those lines, I did your uninstall command, whereas I got this:

Removing Homebrew installation...
Warning: Failed to delete /usr/local/Cellar
rm: /usr/local/Cellar: Permission denied
==> Removing empty directories...
Password:
==> /usr/bin/sudo /usr/bin/find /usr/local/bin /usr/local/etc /usr/local/include /usr/local/lib /usr/local/opt /usr/local/sbin /usr/local/share /usr/local/var /usr/local/Cellar /usr/local/Homebrew /usr/local/Frameworks -name .DS_Store -delete
==> /usr/bin/sudo /usr/bin/find /usr/local/bin /usr/local/etc /usr/local/include /usr/local/lib /usr/local/opt /usr/local/sbin /usr/local/share /usr/local/var /usr/local/Cellar /usr/local/Homebrew /usr/local/Frameworks -depth -type d -empty -exec rmdir {} ;
==> /usr/bin/sudo rmdir /usr/local
rmdir: /usr/local: Operation not permitted
Warning: Failed during: /usr/bin/sudo rmdir /usr/local
Warning: Homebrew partially uninstalled (but there were steps that failed)!
To finish uninstalling rerun this script with `sudo`.
The following possible Homebrew files were not deleted:
/usr/local/Homebrew/
/usr/local/SUPPORTERS.md
/usr/local/etc/
/usr/local/remotedesktop/
/usr/local/share/
/usr/local/var/
You may wish to remove them yourself.

I'm not sure if it meant literally, and was nervous about messing with program files, but I just went and did a search in Finder, found them, and moved to the Trash, so there's nothing in usr/local, but there's various other folders including everything in /bin.  After which 'doctor' gave me errors about unlinked stuff, and the commands to fix, which I did, so now it once again says 'Your system is ready to brew.".  Great.  Config remains the same, and I wasn't sure if I was missing anything else to clean up (should I run a cleanup/upgrade?), so I told it to install telnet again.  Also by the way, when I open a Terminal window to do all this, it includes 'The default interactive shell is now zsh.  To update your account to use zsh, please run `chsh -s /bin/zsh`.' ....  should I do this?

So... I think it's cleaned up and migrated?  Except the main reason I'm doing all this, to use Telnet to port into another server I play on, still won't work.  Previously, I had a bookmark to telnet://theservername:port #  which still spits back this: 
[Command not found: telnet]
[Could not create a new process and open a pseudo-tty.]

Sorry for my very verbose reply and confusion and being able to pin down the answer through various search results... I thought Homebrew would allow me to run that again, and I'm not sure what I need to do next since it all seems properly installed.  Thanks so much for the help!

**Reply by ZhongRuoyu** [reply]
> I wasn't sure if I was missing anything else to clean up (should I run a cleanup/upgrade?)

`brew upgrade` is for upgrading packages. Since you have just installed them (via migration), it should be fine. `brew cleanup` removes stale files, outdated installations, etc. You may run them as needed, but not strictly necessary to solve your problem. See [manpage](https://docs.brew.sh/Manpage) (or run `man brew`) to find out more about these commands (as well as others).

> it includes 'The default interactive shell is now zsh. To update your account to use zsh, please run `chsh -s /bin/zsh`.' .... should I do this?

I assume you are using bash (does `echo $SHELL` give you `/bin/bash`?). You are seeing this because the default shell for macOS has been zsh since Catalina. It's up to you to decide whether to switch to zsh or not.

> to use Telnet to port into another server I play on, still won't work.

Did you run these in your "Next steps":

```
Run these two commands in your terminal to add Homebrew to your PATH:
(echo; echo 'eval "$(/opt/homebrew/bin/brew shellenv)"') >> /Users/emilycollins/.profile
eval "$(/opt/homebrew/bin/brew shellenv)"
```

And if you have done so, did you run `brew install telnet` to install `telnet`? If both your answers are yes: to be sure, are you using the terminal with your default login shell? (If it's the macOS-provided Terminal.app, check Settings > General > Shells open with).

**Comment by emilytweaks** [comment]
Hi, thanks again for your suggestions!
To answer each: yes, I see am using bash.  Just was checking if I needed to update that command suggestion.
I tried to run my Next Steps, copy and pasted each line.  I get no response to either of them.
Yes, I installed telnet, I ended up with this:

==> Downloading https://formulae.brew.sh/api/formula.jws.json
################################################################################################ 100.0%
==> Downloading https://formulae.brew.sh/api/cask.jws.json
################################################################################################ 100.0%
==> Fetching telnet
==> Downloading https://ghcr.io/v2/homebrew/core/telnet/manifests/69
################################################################################################ 100.0%
==> Downloading https://ghcr.io/v2/homebrew/core/telnet/blobs/sha256:59e9f6c98b4a22c314d741d0b54ae3cfb0
################################################################################################ 100.0%
==> Pouring telnet--69.arm64_ventura.bottle.tar.gz
🍺  /opt/homebrew/Cellar/telnet/69: 4 files, 208.1KB
==> Running `brew cleanup telnet`...
Disable this behaviour by setting HOMEBREW_NO_INSTALL_CLEANUP.
Hide these hints with HOMEBREW_NO_ENV_HINTS (see `man brew`).

And I see a telnet file in /opt/homebrew/bin in the Finder, and /opt/homebrew/Cellar/ (that's where the new version should be, right?  Not in /usr which was the old stuff I think?  Sorry, just trying to keep terminology straight.  Then you reminded me to check my settings in my browser of what it was using, and... again yes, opening ssh or telnet tells it to use Terminal.  But I'm still getting this when I try to port to the server through my browser link...

[Command not found: telnet]

[Could not create a new process and open a pseudo-tty.]

Sigh... ok, so telnet just won't work?  Maybe I should try following directions for any of the other clients out there that let you port into Muds/Mucks... just seemed some of them needed Homebrew anyway, and would work through telnet.  Thanks for all the advice.

**Reply by ZhongRuoyu** [reply]
> I tried to run my Next Steps, copy and pasted each line. I get no response to either of them.

Yes, this is expected. It's only to configure your shell environment (so you can run `brew` and the installed software).

> Not in /usr which was the old stuff I think?

`/usr`, except `/usr/local`, is managed by the system. On Intel machines, Homebrew is installed to `/usr/local` by default. On Apple Silicon machines, the default is `/opt/homebrew`. So, this is expected.

> But I'm still getting this when I try to port to the server through my browser link...

So, it seems that you were trying to open a `telnet://` URL from browser? What if you try running `telnet <hostname>` in the terminal? It should work in that case. (Side note: telnet is insecure. Make sure you know what you are doing, and you trust the remote host. Don't perform sensitive actions.)

The underlying reason is that Homebrew's shell environment is not loaded (through `eval "$(/opt/homebrew/bin/brew shellenv)"`) when you open a `telnet` URL through browser.

**Comment by emilytweaks** [comment]
Yes, understood.  
And when I try to simply run the command in a shell instead of 'linking' to it as mentioned, I still get similar errors... 'nodename nor servname provided, or not known'.
And yes, I trust them... it's a muck/mud I've played on for many many years, and since I'm ridiculously simple, I just got used to always using a telnet window to port into them and use plain text.  And to make it even easier, I found that making a bookmark I could click on right from my browser to telnet://hostname:port opened it right up without the need for opening a Terminal window and typing the command/address every single time!  And based off all my searching around before I came here, there were only a few clients that would work with my system anyway, if migration proved to be difficult and I should finally upgrade myself to something less clunky than telnet.  But even some of those didn't seem to be an app (except for Mudlet), but something you built off from existing Homebrew/terminal, such as BeipMu and TinTin++.  So I was stuck asking for help getting Homebrew to work anyway, and wondering how I'd 'link' in to a new client the way I used to, anyhow.  Sounds like I still have some searching/tinkering to do with one of those clients, still not sure why I can't use telnet... but thanks so much for the help with the Homebrew errors!

**Reply by ZhongRuoyu** [reply]
> 'nodename nor servname provided, or not known'

This simply means `telnet` does not recognise the hostname you provided. Note that if you are connecting to a non-default port (the default is 23), you will need to enter `telnet [host] [port]` instead of `telnet [host]:[port]`. (There's a space, not a colon, in between.)

The manual page, accessible through `man telnet`, should give you more information.

> And yes, I trust them...

Pardon me, just wanted to add on to my previous answer: `telnet` is insecure in that anything transmitted from/to the remote host is not encrypted. So please avoid doing anything sensitive.

**Comment by emilytweaks** [comment]
Right, of course.  The lines I gave you were just how I originally learned how to connect through telnet many years ago.  Makes sense that going from a very old system all the way up to Ventura now, it handles/shows/limits some things differently.  So if I try typing it your way...  ... oh hey, it works!  Ha... crazy how just one different symbol changes commands.  It /still/ won't let me likewise tweak it that way through a browser address with telnet:// as I mentioned doing, can't figure why, but that was mainly for my instant-link convenience.

And no, you're fine... I do realize I've stuck with an outdated method, even though I'm careful, so will have to look into proper clients (many of which also require tweaking through Homebrew, so there's that).  At least it's fixed, and thanks.

**Reply by ZhongRuoyu** [reply]
Glad it worked for you.

> It /still/ won't let me likewise tweak it that way through a browser address with telnet:// as I mentioned doing, can't figure why

Because when handling the `telnet://` URL, the terminal only attempts to look for `telnet` under system directories. If it worked on your old Mac, I presume it's because the macOS is old enough to have `telnet` pre-installed (`telnet` was removed from macOS in High Sierra). There are ways to make it work, but I advise against them because they require tinkering the system.

If what you want is a shortcut, perhaps you'll want to try to save the telnet connection in your Terminal.app. Open a terminal window, and press Command-Shift-K to open the "New Remote Connection" window. Select "Remote Login (telnet)" on the left, and click the "+" button below the server list on the right. Enter a name for the remote server, and then click OK. Select the server you've just created, and modify the command below so that it reads `/opt/homebrew/bin/telnet [host] [port]` (just as what you would type in the terminal, but with full path to the `telnet` program.) Next time, you can connect to telnet by pressing Command-Shift-K and selecting the server. If that makes things easier.

**Comment by 615689347** [comment]
Hi

**Comment by emilytweaks** [comment]
Yes, that worked, thanks!  Even though all of this diverted from Homebrew a bit and still need to get with the times and a more modern client.  Of course... the rest makes sense... ha, yes, I finally upgraded from quite an old computer!  Thanks again!
